File: ruby/beer-song/beer_song.rb
class BeerSong
  VERSION = 2

  def verse(num)
    first_line(num) + second_line(num)
  end

  def verses(start, finish)
    start.downto(finish).map do |num|
      first_line(num) + second_line(num)
    end.join("\n")
  end

  def lyrics
    verses(99, 0)
  end

  private

  def first_line(num)
    case num
    when 0
      "No more bottles of beer on the wall, no more bottles of beer.\n"
    when 1
      "1 bottle of beer on the wall, 1 bottle of beer.\n"
    else
      "#{num} bottles of beer on the wall, #{num} bottles of beer.\n"
    end
  end

  def second_line(num)
    case num
    when 0
      "Go to the store and buy some more, 99 bottles of beer on the wall.\n"
    when 1
      "Take it down and pass it around, no more bottles of beer on the wall.\n"
    when 2
      "Take one down and pass it around, 1 bottle of beer on the wall.\n"
    else
      "Take one down and pass it around, #{num-1} bottles of beer on the wall.\n"
    end
  end
end

File: elixir/beer-song/lib/beer_song.ex
defmodule BeerSong do
  @doc """
  Get a single verse of the beer song
  """
  @spec verse(integer()) :: String.t()
  def verse(0) do
    """
    No more bottles of beer on the wall, no more bottles of beer.
    Go to the store and buy some more, 99 bottles of beer on the wall.
    """
  end

  def verse(n) do
    """
    #{bottles(n)} of beer on the wall, #{bottles(n)} of beer.
    Take #{pronoun(n)} down and pass it around, #{bottles(n - 1)} of beer on the wall.
    """
  end

  @doc """
  Get the entire beer song for a given range of numbers of bottles.
  """
  @spec lyrics(Range.t()) :: String.t()
  def lyrics(verses \\ 99..0) do
    verses
    |> Enum.map(&verse/1)
    |> Enum.join("\n")
  end

  defp bottles(0), do: "no more bottles"
  defp bottles(1), do: "1 bottle"
  defp bottles(n), do: "#{n} bottles"

  defp pronoun(1), do: "it"
  defp pronoun(_), do: "one"
end
